server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/UserRoutes.java

```diff
@@ -150,52 +153,47 @@ private String removeUser(Request request, Response response) {
-            throw ErrorResponder.builder()
-                .statusCode(HttpStatus.BAD_REQUEST_400)
-                .type(ErrorType.INVALID_ARGUMENT)
-                .message("Invalid username: it should be between 1 and 255 long without '/'")
-                .cause(e)
-                .haltError();
         } catch (IllegalArgumentException e) {
             LOGGER.info("Invalid user path", e);
             throw ErrorResponder.builder()
                 .statusCode(HttpStatus.BAD_REQUEST_400)
                 .type(ErrorType.INVALID_ARGUMENT)
-                .message("Invalid user path")
+                .message(e.getMessage())
                 .cause(e)
                 .haltError();
         }
     }
 
-    private String upsertUser(Request request, Response response) throws UsersRepositoryException {
+    private HaltException upsertUser(Request request, Response response) throws JsonExtractException {
+        String username = request.params(USER_NAME);
         try {
-            return userService.upsertUser(request.params(USER_NAME),
-                jsonExtractor.parse(request.body()).getPassword(),
-                response);
-        } catch (JsonExtractException e) {
-            LOGGER.info("Error while deserializing addUser request", e);
+            userService.upsertUser(username,
+                jsonExtractor.parse(request.body()).getPassword());
+
+            return halt(HttpStatus.NO_CONTENT_204);
+        } catch (InvalidUsernameException e) {
+            LOGGER.info("Invalid username", e);
             throw ErrorResponder.builder()
                 .statusCode(HttpStatus.BAD_REQUEST_400)
                 .type(ErrorType.INVALID_ARGUMENT)
-                .message("Error while deserializing addUser request")
+                .message(e.getMessage())
                 .cause(e)
                 .haltError();
-        } catch (UserService.InvalidUsername e) {
-            LOGGER.info("Invalid username", e);
+        } catch (UsersRepositoryException e) {
+            String errorMessage = String.format("Error while adding/updating user '%s'", username);
+            LOGGER.info(errorMessage, e);
             throw ErrorResponder.builder()
-                .statusCode(HttpStatus.BAD_REQUEST_400)
-                .type(ErrorType.INVALID_ARGUMENT)
-                .message("Invalid username: it should be between 1 and 255 long without '/'")
+                .statusCode(HttpStatus.INTERNAL_SERVER_ERROR_500)
+                .type(ErrorType.SERVER_ERROR)
+                .message(errorMessage)
                 .cause(e)
                 .haltError();
         } catch (IllegalArgumentException e) {
             LOGGER.info("Invalid user path", e);
             throw ErrorResponder.builder()
                 .statusCode(HttpStatus.BAD_REQUEST_400)
                 .type(ErrorType.INVALID_ARGUMENT)
-                .message("Invalid user path")
+                .message(e.getMessage())
                 .cause(e)
```

Comment:
Won't it be better to rely on default IllegalArgumentException handling here? (remove this catch block)